azure/services/virtualnetworks/virtualnetworks.go

```diff
@@ -150,19 +167,34 @@ func (s *Service) IsManaged(ctx context.Context) (bool, error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.IsManaged")
 	defer done()
 
+	// Report errors if the client is rate limited.
+	if !s.RateLimiter.Reader.TryAccept() {
```

Comment:
This is good for mitigation when we do run into rate limiting. I agree we should try to do this generically across all services (we're already using the retry-after header to requeue create/delete async calls but we need to also use it across all GETs).

However, there is also an aspect of prevention to avoid running into rate limits as much as possible, by reducing the number of calls we make per reconcile (and reducing how often we reconciling which we already do by debouncing successful reconciles thanks to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/pkg/coalescing/reconciler.go).

Low-hanging fruit for this PR would be to remove any extra calls to isVnetManaged since each one makes an API call. The ideal would be to cache the result of isVnetManaged in scope so we only make a single call per AzureCluster reconcile (ie. we don't need to check if it's managed once in the vnet service, AND for every subnet, we can get the result once and then reuse the value for all services, similar to how we cache VM skus in the machine scope.

Reply:
is "cache" in scope as simple as assigning the API result to the scope object instance? (i.e., a new scope object is instantiated in every reconcile loop)

Reply:
could be the API result or even just the boolean value of `isVnetManaged` since that's the only thing we need

Reply:
see https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machine.go#L102 for an example